### ISSUE TITLE
[stack-buffer-overflow](tmo.cpp:78): size `(1 << outbits)` might be too big for array `hdrtoldr`

### ISSUE THREAD

**Issue by benehalo** [thorfdbg/libjpeg#99]
Dear All,

This bug was found on Ubuntu 20.04 64-bit & _libjpeg_ was checked out from the master branch of [GitHub repository](https://github.com/thorfdbg/libjpeg). Its commit is 3d391f06c80b6662d4524d8297e84528b009b774 (Mon Jan 15 07:47:50 2024 +0100).

_libjpeg_ was built with _ASAN_ using _clang-14_. The compile command was:

```bash
cd $PROJECT_SRC
export FLAGS='-DGLOBAL_NEW_IS_OK=1 -DFORTIFY_SOURCE -fstack-protector-all -g -fPIC -fsanitize=address'
make clean 
CC="clang $FLAGS" CXX="clang++ $FLAGS" ./configure
make -j
```

To reproduce:

Download and unzip the attached zip archive, and get PoCs
```bash
$PROJECT_SRC/jpeg -q 10 -R 4 -h [poc]  /dev/null
```

# Bug Analysis

Local variable `hdrtoldr` in function `EncodeC` is a stack array with size 65536
```c
// encodec.cpp:203
UWORD hdrtoldr[65536]; // Its inverse. This table is used to construct the legacy image.
```

Then, the array `hdrtoldr` is passed as the 2'nd argument of function `InvertTable`
```c
// encodec.cpp:363
InvertTable(ldrtohdr,hdrtoldr,8 + hiddenbits,prec);
```

In function `InvertTable`, array `hdrtoldr` (alias `ouput`) is filled with 0 (by invoking `memset`)
```c
void InvertTable(UWORD input[65536],UWORD output[65536],UBYTE inbits,UBYTE outbits)
{ 
  // tmo.cpp:78
  memset(output,0,(1 << outbits) * sizeof(UWORD));
```

 
However, when `(1 << outbits) > 65536`, this invocation of `memset` can result in a stack-buffer-overflow error.

# GDB Analysis
```gdb
Breakpoint 4, EncodeC (source=0x7fffffffe17d "poc-trim/stack-buffer-overflow-0", ldrsource=0x0, target=0x7fffffffe19e "/dev/null", ltable=0x0, quality=10, hdrquality=-1, tabletype=0, residualtt=0, maxerror=0, colortrafo=1, baseline=false, lossless=false, progressive=false, residual=false, optimize=true, accoding=false, rsequential=false, rprogressive=false, raccoding=false, qscan=false, levels=0 '\000', pyramidal=false, writednl=false, restart=0, gamma=0, lsmode=-1, noiseshaping=false, serms=false, losslessdct=false, openloop=false, deadzone=false, lagrangian=false, dering=false, xyz=false, cxyz=false, hiddenbits=4, riddenbits=0, resprec=8, separate=false, median=true, noclamp=false, smooth=0, dctbypass=false, sub=0x0, ressub=0x0, alpha=0x0, alphamode=1, matte_r=0, matte_g=0, matte_b=0, alpharesiduals=false, alphaquality=70, alphahdrquality=0, alphatt=0, residualalphatt=0, ahiddenbits=0, ariddenbits=0, aresprec=8, aopenloop=false, adeadzone=false, alagrangian=false, adering=false, aserms=false, abypass=false, quantsteps=0x0, residualquantsteps=0x0, alphasteps=0x0, residualalphasteps=0x0) at encodec.cpp:203
203       UWORD hdrtoldr[65536]; // Its inverse. This table is used to construct the legacy image.
(gdb) c
Continuing.

Breakpoint 1, EncodeC (source=0x7fffffffe17d "poc-trim/stack-buffer-overflow-0", ldrsource=0x0, target=0x7fffffffe19e "/dev/null", ltable=0x0, quality=10, hdrquality=-1, tabletype=0, residualtt=0, maxerror=0, colortrafo=1, baseline=false, lossless=false, progressive=false, residual=false, optimize=true, accoding=false, rsequential=false, rprogressive=false, raccoding=false, qscan=false, levels=0 '\000', pyramidal=false, writednl=false, restart=0, gamma=0, lsmode=-1, noiseshaping=false, serms=false, losslessdct=false, openloop=false, deadzone=false, lagrangian=false, dering=false, xyz=false, cxyz=false, hiddenbits=4, riddenbits=0, resprec=8, separate=false, median=true, noclamp=false, smooth=0, dctbypass=false, sub=0x0, ressub=0x0, alpha=0x0, alphamode=1, matte_r=0, matte_g=0, matte_b=0, alpharesiduals=false, alphaquality=70, alphahdrquality=0, alphatt=0, residualalphatt=0, ahiddenbits=0, ariddenbits=0, aresprec=8, aopenloop=false, adeadzone=false, alagrangian=false, adering=false, aserms=false, abypass=false, quantsteps=0x0, residualquantsteps=0x0, alphasteps=0x0, residualalphasteps=0x0) at encodec.cpp:363
363             InvertTable(ldrtohdr,hdrtoldr,8 + hiddenbits,prec);
(gdb) c
Continuing.

Breakpoint 2, InvertTable (input=0x7ffffff3b590, output=0x7ffffffbb990, inbits=12 '\f', outbits=20 '\024') at tmo.cpp:78
78        memset(output,0,(1 << outbits) * sizeof(UWORD));
(gdb) p 1 << outbits
$1 = 1048576
(gdb) p (1 << outbits) > 65536
$2 = true
(gdb) bt
#0  InvertTable (input=0x7ffffff3b590, output=0x7ffffffbb990, inbits=12 '\f', outbits=20 '\024') at tmo.cpp:78
#1  0x00000000005164ac in EncodeC (source=0x7fffffffe17d "poc-trim/stack-buffer-overflow-0", ldrsource=0x0, target=0x7fffffffe19e "/dev/null", ltable=0x0, quality=10, hdrquality=-1, tabletype=0, residualtt=0, maxerror=0, colortrafo=1, baseline=false, lossless=false, progressive=false, residual=false, optimize=true, accoding=false, rsequential=false, rprogressive=false, raccoding=false, qscan=false, levels=0 '\000', pyramidal=false, writednl=false, restart=0, gamma=0, lsmode=-1, noiseshaping=false, serms=false, losslessdct=false, openloop=false, deadzone=false, lagrangian=false, dering=false, xyz=false, cxyz=false, hiddenbits=4, riddenbits=0, resprec=8, separate=false, median=true, noclamp=false, smooth=0, dctbypass=false, sub=0x0, ressub=0x0, alpha=0x0, alphamode=1, matte_r=0, matte_g=0, matte_b=0, alpharesiduals=false, alphaquality=70, alphahdrquality=0, alphatt=0, residualalphatt=0, ahiddenbits=0, ariddenbits=0, aresprec=8, aopenloop=false, adeadzone=false, alagrangian=false, adering=false, aserms=false, abypass=false, quantsteps=0x0, residualquantsteps=0x0, alphasteps=0x0, residualalphasteps=0x0) at encodec.cpp:363
#2  0x00000000005022ea in main (argc=3, argv=0x7fffffffde60) at main.cpp:760
```

# ASAN says
```gdb
=================================================================
==785767==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffee24a8710 at pc 0x0000004c110c bp 0x7ffee2406cf0 sp 0x7ffee24064a0
WRITE of size 2097152 at 0x7ffee24a8710 thread T0
    #0 0x4c110b in __asan_memset /llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:26:3
    #1 0x50d504 in InvertTable(unsigned short*, unsigned short*, unsigned char, unsigned char) /data/symccgo/bug/libjpeg/libjpeg/cmd/tmo.cpp:78:3
    #2 0x5164ab in EncodeC(char const*, char const*, char const*, char const*, int, int, int, int, int, int, bool, bool, bool, bool, bool, bool, bool, bool, bool, bool, unsigned char, bool, bool, unsigned int, double, int, bool, bool, bool, bool, bool, bool, bool, bool, bool, int, int, int, bool, bool, bool, int, bool, char const*, char const*, char const*, int, int, int, int, bool, int, int, int, int, int, int, int, bool, bool, bool, bool, bool, bool, char const*, char const*, char const*, char const*) /data/symccgo/bug/libjpeg/libjpeg/cmd/encodec.cpp:363:9
    #3 0x5022e9 in main /data/symccgo/bug/libjpeg/libjpeg/cmd/main.cpp:760:7
    #4 0x7f5663813082 in __libc_start_main /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:308:16
    #5 0x41c43d in _start (/data/symccgo/bug/libjpeg/obj-asan-dbg/jpeg+0x41c43d)

Address 0x7ffee24a8710 is located in stack of thread T0 at offset 658032 in frame
    #0 0x51294f in EncodeC(char const*, char const*, char const*, char const*, int, int, int, int, int, int, bool, bool, bool, bool, bool, bool, bool, bool, bool, bool, unsigned char, bool, bool, unsigned int, double, int, bool, bool, bool, bool, bool, bool, bool, bool, bool, int, int, int, bool, bool, bool, int, bool, char const*, char const*, char const*, int, int, int, int, bool, int, int, int, int, int, int, int, bool, bool, bool, bool, bool, bool, char const*, char const*, char const*, char const*) /data/symccgo/bug/libjpeg/libjpeg/cmd/encodec.cpp:95

  This frame has 61 object(s):
    [32, 36) 'ahiddenbits.addr'
    [48, 112) 'pscan1' (line 96)
    [144, 224) 'pscan2' (line 102)
    [256, 336) 'pscan3' (line 109)
    [368, 448) 'pscan4' (line 116)
    [480, 576) 'pscan5' (line 123)
    [608, 688) 'pscan6' (line 131)
    [720, 800) 'pscan7' (line 138)
    [832, 896) 'qscan1' (line 146)
    [928, 992) 'qscan2' (line 152)
    [1024, 1072) 'qscan3' (line 158)
    [1104, 1136) 'rscan1' (line 164)
    [1168, 1216) 'rscan2' (line 168)
    [1248, 1296) 'rscan3' (line 173)
    [1328, 1376) 'rscan4' (line 178)
    [1408, 1456) 'rscan5' (line 183)
    [1488, 1536) 'rscan6' (line 188)
    [1568, 1616) 'rscan7' (line 193)
    [1648, 132720) 'ldrtohdr' (line 201)
    [132976, 264048) 'red' (line 202)
    [264304, 395376) 'green' (line 202)
    [395632, 526704) 'blue' (line 202)
    [526960, 658032) 'hdrtoldr' (line 203)
    [658288, 789360) 'alphaldrtohdr' (line 204) <== Memory access at offset 658032 partially underflows this variable
    [789616, 789620) 'subx' (line 206) <== Memory access at offset 658032 partially underflows this variable
    [789632, 789636) 'suby' (line 206) <== Memory access at offset 658032 partially underflows this variable
    [789648, 789652) 'ressubx' (line 207) <== Memory access at offset 658032 partially underflows this variable
    [789664, 789668) 'ressuby' (line 207) <== Memory access at offset 658032 partially underflows this variable
    [789680, 789936) 'qntmatrix' (line 208) <== Memory access at offset 658032 partially underflows this variable
    [790000, 790256) 'qntmatrixchroma' (line 208) <== Memory access at offset 658032 partially underflows this variable
    [790320, 790576) 'residualmatrix' (line 209) <== Memory access at offset 658032 partially underflows this variable
    [790640, 790896) 'residualmatrixchroma' (line 209) <== Memory access at offset 658032 partially underflows this variable
    [790960, 791216) 'alphamatrix' (line 210) <== Memory access at offset 658032 partially underflows this variable
    [791280, 791536) 'alphamatrixchroma' (line 210) <== Memory access at offset 658032 partially underflows this variable
    [791600, 791856) 'residualalphamatrix' (line 211) <== Memory access at offset 658032 partially underflows this variable
    [791920, 792176) 'residualalphamatrixchroma' (line 211) <== Memory access at offset 658032 partially underflows this variable
    [792240, 792244) 'width' (line 242) <== Memory access at offset 658032 partially underflows this variable
    [792256, 792260) 'height' (line 242) <== Memory access at offset 658032 partially underflows this variable
    [792272, 792276) 'depth' (line 242) <== Memory access at offset 658032 partially underflows this variable
    [792288, 792292) 'prec' (line 242) <== Memory access at offset 658032 partially underflows this variable
    [792304, 792308) 'alphaprec' (line 243) <== Memory access at offset 658032 partially underflows this variable
    [792320, 792321) 'flt' (line 244) <== Memory access at offset 658032 partially underflows this variable
    [792336, 792337) 'alphaflt' (line 245) <== Memory access at offset 658032 partially underflows this variable
    [792352, 792353) 'big' (line 246) <== Memory access at offset 658032 partially underflows this variable
    [792368, 792369) 'alphabig' (line 247) <== Memory access at offset 658032 partially underflows this variable
    [792384, 792385) 'fullrange' (line 248) <== Memory access at offset 658032 partially underflows this variable
    [792400, 792404) 'ldrdepth' (line 254) <== Memory access at offset 658032 partially underflows this variable
    [792416, 792420) 'ldrwidth' (line 255) <== Memory access at offset 658032 partially underflows this variable
    [792432, 792436) 'ldrheight' (line 256) <== Memory access at offset 658032 partially underflows this variable
    [792448, 792452) 'ldrprec' (line 257) <== Memory access at offset 658032 partially underflows this variable
    [792464, 792465) 'ldrflt' (line 258) <== Memory access at offset 658032 partially underflows this variable
    [792480, 792481) 'ldrbig' (line 259) <== Memory access at offset 658032 partially underflows this variable
    [792496, 792632) 'bmm' (line 426) <== Memory access at offset 658032 partially underflows this variable
    [792704, 792728) 'bmhook' (line 427) <== Memory access at offset 658032 partially underflows this variable
    [792768, 792792) 'ldrhook' (line 428) <== Memory access at offset 658032 partially underflows this variable
    [792832, 792856) 'alphahook' (line 429) <== Memory access at offset 658032 partially underflows this variable
    [792896, 793648) 'alphatags' (line 432) <== Memory access at offset 658032 partially underflows this variable
    [793776, 795264) 'tags' (line 503) <== Memory access at offset 658032 partially underflows this variable
    [795392, 795416) 'filehook' (line 696) <== Memory access at offset 658032 partially underflows this variable
    [795456, 795504) 'iotags' (line 697) <== Memory access at offset 658032 partially underflows this variable
    [795536, 795544) 'error' (line 732) <== Memory access at offset 658032 partially underflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.cpp:26:3 in __asan_memset
Shadow bytes around the buggy address:
  0x10005c48d090: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d0a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d0b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d0c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d0d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10005c48d0e0: 00 00[f2]f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x10005c48d0f0: f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2 f2
  0x10005c48d100: f2 f2 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d110: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d120: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10005c48d130: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==785767==ABORTING
=====
```

# PoC
[poc-sbo.zip](https://github.com/thorfdbg/libjpeg/files/14730974/poc-sbo.zip)

**Comment by thorfdbg** [comment]
This is also an garbadge-in garbadge-out problem. The ppm file passed in has an invalid precision defined.

**Comment by thorfdbg** [comment]
This is due to an invalid ppm. Added a bit of input validation for ppm, fixed in 1.68.
